library/std/src/sys/unix/fd.rs

```diff
@@ -284,6 +284,10 @@ impl<'a> Read for &'a FileDesc {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
         (**self).read(buf)
     }
+
+    fn read_buf(&mut self, cursor: BorrowedCursor<'_>) -> io::Result<()> {
+        (**self).read_buf(cursor)
+    }
```

Comment:
BTW, should implementation of other methods be moved here and delegation removed to avoid this kind of issues?

Reply:
Why does this impl on `&FileDesc` exist at all?

Reply:
For `read_to_end` apparently. With impl removed:

```console
warning: function cannot return without recursing
   --> library/std/src/sys/unix/fd.rs:112:5
    |
112 |     pub fn read_to_end(&self, buf: &mut Vec<u8>) -> io::Result<usize> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
113 |         let mut me = self;
114 |         (&mut me).read_to_end(buf)
    |         -------------------------- recursive call site

```

Reply:
Apologies, I removed my comment (asking why the impl exists) immediately after posting it.

I had meant why was the impl on `&FileDesc` when I (incorrectly) assumed it was delegating to an impl of the same trait on `FileDesc`—and I think the trait isn't directly implemented on `FileDesc` is to permit reading behind a shared reference.

Reply:
Hrm, there are a lot of other `Read` impls related to stdio and process IO. Wouldn't those suffer from the same perf issues?

Reply:
In general implementations should probably forward `read`, `read_buf`, `read_vectored`, and `is_read_vectored`, which isn't consistently the case right now. I can take a look into this in a followup, or here if preferred.